[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113679

Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

fegin committed:
[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState

Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

[ghstack-poisoned]

fegin committed:
Update on "[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState"


Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

[ghstack-poisoned]

fegin committed:
Update on "[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState"


Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/replicate.py:
```diff
@@ -15,83 +14,85 @@
     def __init__(self) -> None:
         super().__init__()
-        self.module: Optional[nn.Module] = None
+        self.module: nn.Module = nn.ParameterList()
```
Is the `nn.ParameterList()` just a dummy so that we can type `self.module` as `nn.Module`?

fegin replied:
Yes, I should add a comment about this.

fegin committed:
Update on "[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState"


Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

cc H-Huang awgu kwen2501 wanchaol fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

fegin committed:
Update on "[replicate] Simplify replicate() init logic and remove unnecessary variables in _ReplicateState"


Many variables _ReplicateState are created because replicate() was lazy initialized. This PR removes these variables and simplifes the logic.y

Differential Revision: [D51317874](https://our.internmc.facebook.com/intern/diff/D51317874/)

cc H-Huang awgu kwen2501 wanchaol fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]
